packages/pipeline-editor/src/validation.spec.ts

```diff
@@ -0,0 +1,343 @@
+  {
+    it: 'should detect joined cycles',
+    given: [
+      { id: 'l0', srcNodeId: 'a', trgNodeId: 'b', type: 'nodeLink' },
+      { id: 'l1', srcNodeId: 'b', trgNodeId: 'a', type: 'nodeLink' },
+      { id: 'l2', srcNodeId: 'b', trgNodeId: 'c', type: 'nodeLink' },
+      { id: 'l3', srcNodeId: 'c', trgNodeId: 'b', type: 'nodeLink' }
+    ],
+    expected: ['l0', 'l1', 'l2', 'l3']
+  },
+
+  // ╭───╮      ╭───╮      ╭───╮
+  // │ a │──l0─►│ b │──l2─►│ c │
+  // ╰───╯      ╰───╯      ╰───╯
+  //   ▲         │ ▲         │
+  //   ╰────l1───╯ ╰───l3────╯
+  {
+    it: 'should detect joined cycles - shuffled',
+    given: [
+      { id: 'l3', srcNodeId: 'c', trgNodeId: 'b', type: 'nodeLink' },
+      { id: 'l1', srcNodeId: 'b', trgNodeId: 'a', type: 'nodeLink' },
+      { id: 'l2', srcNodeId: 'b', trgNodeId: 'c', type: 'nodeLink' },
+      { id: 'l0', srcNodeId: 'a', trgNodeId: 'b', type: 'nodeLink' }
+    ],
+    expected: ['l0', 'l1', 'l2', 'l3']
+  },
+
+  // ╭───╮      ╭───╮
+  // │ a │──l0─►│ b │
+  // ╰───╯      ╰───╯
+  //   │          │
+  //  l3         l1
+  //   ▼          ▼
+  // ╭───╮      ╭───╮
+  // │ d │◄─l2──│ c │
+  // ╰───╯      ╰───╯
+  {
+    it: 'should not detect fake cycle',
+    given: [
+      { id: 'l0', srcNodeId: 'a', trgNodeId: 'b', type: 'nodeLink' },
+      { id: 'l1', srcNodeId: 'b', trgNodeId: 'c', type: 'nodeLink' },
+      { id: 'l2', srcNodeId: 'c', trgNodeId: 'd', type: 'nodeLink' },
+      { id: 'l3', srcNodeId: 'a', trgNodeId: 'd', type: 'nodeLink' }
+    ],
+    expected: []
+  },
+
+  //   ╭────l4────╮
+  //   ▼          │
+  // ╭───╮      ╭───╮
+  // │ a │──l0─►│ b │
+  // ╰───╯      ╰───╯
+  //   │          ▲
+  //  l2         l1
+  //   ▼          │
+  // ╭───╮      ╭───╮
+  // │ d │◄─l3──│ c │
+  // ╰───╯      ╰───╯
+  {
+    it: 'should detect cycle when a fork that gets checked first is safe',
+    given: [
+      { id: 'l0', srcNodeId: 'a', trgNodeId: 'b', type: 'nodeLink' },
+      { id: 'l1', srcNodeId: 'c', trgNodeId: 'b', type: 'nodeLink' },
+      { id: 'l2', srcNodeId: 'a', trgNodeId: 'd', type: 'nodeLink' },
+      { id: 'l3', srcNodeId: 'c', trgNodeId: 'd', type: 'nodeLink' },
+      { id: 'l4', srcNodeId: 'b', trgNodeId: 'a', type: 'nodeLink' }
+    ],
+    expected: ['l0', 'l4']
+  },
+
+  // ╭───╮      ╭───╮      ╭───╮      ╭───╮      ╭───╮
+  // │ a │──l0─►│ b │──l1─►│ c │──l2─►│ d │──l3─►│ e │
+  // ╰───╯      ╰───╯      ╰───╯      ╰───╯      ╰───╯
+  //                         │
+  //                        l4
+  //                         ▼
+  //                       ╭───╮      ╭───╮      ╭───╮      ╭───╮      ╭───╮
+  //                       │ f │──l5─►│ g │──l6─►│ h │──l7─►│ i │──l8─►│ j │
+  //                       ╰───╯      ╰───╯      ╰───╯      ╰───╯      ╰───╯
+  //                                                          ▲          │
+  //                                                          ╰────l9────╯
+  {
+    it: 'should handle long forks',
```

Comment:
for the algorithm, what is the difference between this scenario and the one described on the test below?

Reply:
When the algorithm gets to node `C` it either traverses down `l2` or `l4` first depending on the order of the links, this just checks that it doesn't matter which it traverses first